File: 98-ValidateBinarySearchTree/98-ValidateBinarySearchTree.cpp
// Last updated: 6/11/2026, 11:36:48 AM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, long long minv, long long maxv){
        if(root == nullptr){
            return true;
        }
        if(root->val>minv && root->val<maxv){
            return helper(root->left, minv, root->val) && helper(root->right, root->val, maxv);
        }
        return false;
    }

    bool isValidBST(TreeNode* root) {
        return helper(root, -1e18, 1e18);
    }

};

File: 98-ValidateBinarySearchTree/98-ValidateBinarySearchTree.py
# Last updated: 6/11/2026, 11:36:53 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def isValidBST(self, root: Optional[TreeNode]) -> bool:
    #     self.isBSTHelper(root,float('-inf'), float('inf'))
    # def isBSTHelper(self, node, lower, high):
    #     if not node:
    #         return True

    #     if not (lower<node.val and node.val<high):
    #         return False

    #     return self.isBSTHelper(node.left, lower, node.val) and self.isBSTHelper(node.right, node.val, high)

    def isValidBST(self, root: TreeNode) -> bool:
        def valid(node, left, right):
            if not node:
                return True
            if not(left<node.val<right):
                return False
            # if(left > node.val or node.val<right):
            #     return False

            return valid(node.left, left, node.val) and valid(
                node.right, node.val, right
            )

        return valid(root, float("-inf"), float("inf"))